spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkScan.java

```diff
@@ -193,15 +200,39 @@ public boolean supportColumnarReads(InputPartition partition) {
     }
   }
 
+  static long numFilesToScan(CombinedScanTask scanTask) {
+    long fileCount = 0L;
+    for (FileScanTask file : scanTask.files()) {
+      fileCount += 1L;
+    }
+    return fileCount;
+  }
+
   private static class RowReader extends RowDataReader implements PartitionReader<InternalRow> {
+    private long numFilesToRead;
+
     RowReader(ReadTask task) {
       super(task.task, task.table(), task.expectedSchema(), task.isCaseSensitive());
+      numFilesToRead = numFilesToScan(task.task);
+    }
+
+    @Override
+    public CustomTaskMetric[] currentMetricsValues() {
+      return new CustomTaskMetric[] { new TaskNumFiles(numFilesToRead) };
     }
   }
 
   private static class BatchReader extends BatchDataReader implements PartitionReader<ColumnarBatch> {
+    private long numFilesToRead;
+
     BatchReader(ReadTask task, int batchSize) {
       super(task.task, task.table(), task.expectedSchema(), task.isCaseSensitive(), batchSize);
+      numFilesToRead = numFilesToScan(task.task);
+    }
+
+    @Override
+    public CustomTaskMetric[] currentMetricsValues() {
+      return new CustomTaskMetric[] { new TaskNumFiles(numFilesToRead) };
     }
```

Comment:
Note: Spark calls this every 100 rows for each `PartitionReader`. The `numFilesToRead` can be computed up front and stored, and is. I wonder if it is worth caching the `CustomTaskMetric[]` as well.